pkg/go/transformer/jsontodsl.go

```diff
@@ -10,6 +10,59 @@ import (
 	"google.golang.org/protobuf/encoding/protojson"
 )
 
+type DirectAssignmentValidator struct {
+	occurred   int
+	stateStack []pb.Userset
+}
+
+func (v *DirectAssignmentValidator) incr() {
+	v.occurred++
+}
+
+func (v *DirectAssignmentValidator) occurrences() int {
+	return v.occurred
+}
+
+func (v *DirectAssignmentValidator) reset() {
+	v.stateStack = []pb.Userset{}
+	v.occurred = 0
+}
+
+var validator = DirectAssignmentValidator{
```

Comment:
Similar to the JS implementation, I think we should not be using a global variable and rely instead on an instantiated instance that's passed around for the duration of the validation.

However - considering how excited I am to ship this feature and have all the tests passing, let's defer that to another PR.

I would love to chat w/ both you and @d-jeffery if the global variable is required for some reason that I'm not considering.

Reply:
@rhamzeh that makes sense to me; I wanted to start with a fairly direct implementation of the existing JS implementation to get something out the door, then I think we can refactor both to a non-global. Want me to create an issue for that?